flask_stupe/pagination.py

```diff
@@ -8,10 +8,35 @@
 
 
 if pymongo:
-    def _paginate(cursor, skip=None, limit=None, sort=None, count=True):
-        metadata = getattr(request, "metadata", None)
-        if count and isinstance(metadata, dict):
-            metadata.update(count=cursor.count())
+
+    def _get_link_header(limit, skip, total_count=None):
+        template = '<{}?limit={}&skip={{skip}}>; rel="{{rel}}"'
+        template = template.format(request.base_url, limit)
+        links = [
+            template.format(skip=0, rel="first"),
+            template.format(skip=skip - limit, rel="prev"),
+            template.format(skip=skip + limit, rel="next")
+        ]
+        if total_count:
+            links.append(template.format(skip=total_count - limit, rel="last"))
+        return ", ".join(links)
+
+    def _paginate(cursor, skip=None, limit=None, sort=None, count=True,
+                  headers=False):
+        if headers:
+            headers = getattr(request, "response_headers", None)
+            if isinstance(headers, dict):
+                total_count = None
+                if count:
+                    total_count = cursor.count()
+                    headers["X-Total-Count"] = total_count
+                if limit is not None and skip is not None:
+                    headers["Link"] = _get_link_header(limit, skip,
+                                                       total_count)
+        else:
+            metadata = getattr(request, "metadata", None)
+            if count and isinstance(metadata, dict):
+                metadata.update(count=cursor.count())
```

Comment:
Honestly I don't really like this whole block, but this is probably good enough for a v1.

I think that a better approach would be to have the information gathered once for all at the beginning, and then apply it to whatever place it should be applied; i.e. generate the relatives URLs (and the total-count if requested) as a first step, and then apply them to the metadatas or headers according to the `headers` field in a second step.

The best part of it would the that we would have the same feature set whether the user choose to use headers or to encapsulate the JSON response.

Reply:
Or we could just remove the encapsulation thing, as discussed in [the other thread](https://github.com/numberly/flask-stupe/pull/27/files#r237005351).

Reply:
I like this!

Reply:
Though I'd suggest for this PR to only have the links for the header "mode", so we don't need to code compatibility with `request.metadata` to then remove it in the PR that will drop said encapsulation.

Reply:
As discussed IRL, let's keep the encapsulation as it's needed for list anyway, and put the links in both the headers and encapsulation.

Reply:
Seen with you: we can't remove totally the encapsulation otherwise we would return lists as top level JSON and that is not standard/dirty.
So if we stick with encapsulation, but still want headers, let's make it a default and have it in both places :)